python/singa/autograd.py

```diff
@@ -3239,176 +3238,152 @@ def __init__(self,
-            batchsize = xs[0].shape[0]
-            out = []
-            h, c = self.step_forward(xs[0], h0, c0, self.Wx, self.Wh, self.Bx,
+        # xs: a tuple or list of input tensors
+        # h0_c0: a tuple of (h0, c0)
+        h0, c0 = h0_c0
+        if not isinstance(xs, list):
+            xs = list(xs)
+        inputs = xs + list((h0, c0))
+        self.device_check(*inputs)
+        # self.device_check(inputs[0], *self.params)
+        self.device_check(inputs[0],
+                          *(self.Wx + self.Wh + self.Bx + self.Bh))
+        batchsize = xs[0].shape[0]
+        out = []
+        h, c = self.step_forward(xs[0], h0, c0, self.Wx, self.Wh, self.Bx,
+                                 self.Bh)
+        out.append(h)
+        for x in xs[1:]:
+            assert x.shape[0] == batchsize
+            h, c = self.step_forward(x, h, c, self.Wx, self.Wh, self.Bx,
                                      self.Bh)
             out.append(h)
-            for x in xs[1:]:
-                assert x.shape[0] == batchsize
-                h, c = self.step_forward(x, h, c, self.Wx, self.Wh, self.Bx,
-                                         self.Bh)
-                out.append(h)
-            return out, h, c
-        elif self.backend == "cudnn":
-            if not singa.USE_CUDA:
-                raise Exception("Could not use cudnn without cuda compiled.\n")
+        return out, h, c
 
-            cpp_x = singa.VecTensor()
-            [cpp_x.append(i.data) for i in xs]
+    def step_forward(self, x, h, c, Wx, Wh, Bx, Bh):
+        y1 = matmul(x, Wx[0])
+        y1 = add_bias(y1, Bx[0], axis=0)
+        y2 = matmul(h, Wh[0])
+        y2 = add_bias(y2, Bh[0], axis=0)
+        i = add(y1, y2)
+        i = sigmoid(i)
+
+        y1 = matmul(x, Wx[1])
+        y1 = add_bias(y1, Bx[1], axis=0)
+        y2 = matmul(h, Wh[1])
+        y2 = add_bias(y2, Bh[1], axis=0)
+        f = add(y1, y2)
+        f = sigmoid(f)
+
+        y1 = matmul(x, Wx[2])
+        y1 = add_bias(y1, Bx[2], axis=0)
+        y2 = matmul(h, Wh[2])
+        y2 = add_bias(y2, Bh[2], axis=0)
+        o = add(y1, y2)
+        o = sigmoid(o)
+
+        y1 = matmul(x, Wx[3])
+        y1 = add_bias(y1, Bx[3], axis=0)
+        y2 = matmul(h, Wh[3])
+        y2 = add_bias(y2, Bh[3], axis=0)
+        g = add(y1, y2)
+        g = tanh(g)
+
+        cout1 = mul(f, c)
+        cout2 = mul(i, g)
+        cout = add(cout1, cout2)
+
+        hout = tanh(cout)
+        hout = mul(o, hout)
+        return hout, cout
+
+
+class _RNN_cudnn(Operation):
+    def __init__(self, handle):
+        super(_RNN_cudnn, self).__init__()
+        self.handle = handle
 
-            self.W = Tensor(shape=(self.rnn_handle.weights_size,),
-                            requires_grad=True,
-                            stores_grad=True)
-            self.W.gaussian(0.0, 1.0)
+    def forward(self, x, W):
+        y = singa.GpuRNNForwardTraining(x, W, self.handle)
```

Comment:
if training is false, need to call GpuRNNForwardInference?

Reply:
this is resolved